plugins/Toolbox/src/CloudSync/DownloadPresenter.py

```diff
@@ -0,0 +1,127 @@
+
+
+## Downloads a set of packages from the Ultimaker Cloud Marketplace
+# use download() exactly once: should not be used for multiple sets of downloads since this class contains state
+class DownloadPresenter:
+
+    def __init__(self, app: CuraApplication):
+        # Emits (Dict[str, str], List[str]) # (success_items, error_items)
+        # Dict{success_package_id, temp_file_path}
+        # List[errored_package_id]
+        self.done = Signal()
+
+        self._app = app
+        self._scope = UltimakerCloudScope(app)
+
+        self._started = False
+        self._progress_message = self._createProgressMessage()
+        self._progress = {}  # type: Dict[str, Dict[str, Any]] # package_id, Dict
+        self._error = []  # type: List[str] # package_id
+
+    def download(self, model: SubscribedPackagesModel):
+        if self._started:
+            Logger.error("Download already started. Create a new %s instead", self.__class__.__name__)
+            return
+
+        manager = HttpRequestManager.getInstance()
+        for item in model.items:
+            package_id = item["package_id"]
+
+            request_data = manager.get(
+                item["download_url"],
+                callback = lambda reply, pid = package_id: self._onFinished(pid, reply),
+                download_progress_callback = lambda rx, rt, pid = package_id: self._onProgress(pid, rx, rt),
+                error_callback = lambda rx, rt, pid = package_id: self._onProgress(pid, rx, rt),
+                scope = self._scope)
+
+            self._progress[package_id] = {
+                "received": 0,
+                "total": 1,  # make sure this is not considered done yet. Also divByZero-safe
+                "file_written": None,
+                "request_data": request_data
+            }
+
+        self._started = True
+        self._progress_message.show()
+
+    def abort(self):
+        manager = HttpRequestManager.getInstance()
+        for item in self._progress.values():
+            manager.abortRequest(item["request_data"])
+
+    # Aborts all current operations and returns a copy with the same settings such as app and scope
+    def resetCopy(self):
+        self.abort()
+        self.done.disconnectAll()
+        return DownloadPresenter(self._app)
+
+    def _createProgressMessage(self) -> Message:
+        return Message(i18n_catalog.i18nc(
+            "@info:generic",
+            "\nSyncing..."),
+            lifetime = 0,
+            use_inactivity_timer=False,
+            progress = 0.0,
+            title = i18n_catalog.i18nc("@info:title", "Changes detected from your Ultimaker account", ))
+
+    def _onFinished(self, package_id: str, reply: QNetworkReply):
+        self._progress[package_id]["received"] = self._progress[package_id]["total"]
+
+        try:
+            with tempfile.NamedTemporaryFile(mode ="wb+", suffix =".curapackage", delete = False) as temp_file:
```

Comment:
If `... deleted = False) ...`, where _does_ it get deleted?

Reply:
SyncOrchestrator._onLicencseAnswers